src/python/pants/util/frozendict.py

```diff
@@ -87,6 +87,27 @@ def __hash__(self) -> int:
     def __repr__(self) -> str:
         return f"{type(self).__name__}({self._data!r})"
 
+    @classmethod
+    def freeze_json_dict(cls, d: dict[Any, Any]) -> FrozenDict[str, Any]:
+        result = {}
+        for k, v in d.items():
+            if not isinstance(k, str):
+                raise AssertionError("Got non-`str` key for FrozenDict.freeze_json_dict.")
+
+            f: Any = None
+            if isinstance(v, list):
+                f = tuple(v)
+            elif isinstance(v, dict):
+                f = cls.freeze_json_dict(v)
+            elif isinstance(v, str) or isinstance(v, int):
+                f = v
+            else:
+                raise AssertionError(
+                    f"Unsupported value type for FrozenDict.freeze_json_dict: {type(v)}"
+                )
+            result[k] = f
+        return cls(result)
+
```

Comment:
This just landed:

https://github.com/pantsbuild/pants/blob/adea9936f7444a4090fe829c2264497be99d9f89/src/python/pants/util/frozendict.py#L51-L65